src/serverenvironment.cpp

```diff
@@ -2199,3 +2211,78 @@ bool ServerEnvironment::migratePlayersDatabase(const GameParams &game_params,
 	}
 	return true;
 }
+
+AuthDatabase *ServerEnvironment::openAuthDatabase(
+		const std::string &name, const std::string &savedir, const Settings &conf)
+{
+	if (name == "sqlite3")
+		return new AuthDatabaseSQLite3(savedir);
+
+	if (name == "files")
+		return new AuthDatabaseFiles(savedir);
+
+	throw BaseException(std::string("Database backend ") + name + " not supported.");
+}
+
+bool ServerEnvironment::migrateAuthDatabase(
+		const GameParams &game_params, const Settings &cmd_args)
+{
+	std::string migrate_to = cmd_args.get("migrate-auth");
+	Settings world_mt;
+	std::string world_mt_path = game_params.world_path + DIR_DELIM + "world.mt";
+	if (!world_mt.readConfigFile(world_mt_path.c_str())) {
+		errorstream << "Cannot read world.mt!" << std::endl;
+		return false;
+	}
+
+	if (!world_mt.exists("auth_backend")) {
+		errorstream << "Please specify your current backend in world.mt:"
+			    << std::endl
+			    << "	auth_backend = {files|sqlite3}" << std::endl;
+		return false;
+	}
+
+	std::string backend = world_mt.get("auth_backend");
+	if (backend == migrate_to) {
+		errorstream << "Cannot migrate: new backend is same"
+			    << " as the old one" << std::endl;
+		return false;
+	}
+
+	try {
+		AuthDatabase *srcdb = ServerEnvironment::openAuthDatabase(
```

Comment:
can you store AuthDatabase pointers in a unique_ptr please ?
There are some possible leaks here when exceptions are triggered

Reply:
Good point. Do you mean only in the `migrate` function, or in general?

Reply:
At least here as it's a single usage, but it can be used in the server part too i think

Reply:
Fixed this at least. Will have to see about the other parts.